arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/map.kt

```diff
@@ -321,227 +317,281 @@ public inline fun <K, E, A, B> Map<K, A>.mapOrAccumulate(
-public fun <K, A, B, C> Map<K, C>.unalign(fa: (Map.Entry<K, C>) -> Ior<A, B>): Pair<Map<K, A>, Map<K, B>> =
-  mapValues(fa).unalign()
+public inline fun <K, A, B, C> Map<K, C>.unalign(fa: (Map.Entry<K, C>) -> Ior<A, B>): Pair<Map<K, A>, Map<K, B>> {
+  val lefts = mutableMapOf<K, A>()
+  val rights = mutableMapOf<K, B>()
+  forEach { entry ->
+    fa(entry).fold(
+      { lefts[entry.key] = it },
+      { rights[entry.key] = it },
+      { a, b ->
+        lefts[entry.key] = a
+        rights[entry.key] = b
+      }
+    )
+  }
+  return lefts to rights
+}
 
 /**
  * Unzips the structure holding the resulting elements in an `Pair`
  *
  * ```kotlin
  * import arrow.core.*
+ * import io.kotest.matchers.shouldBe
  *
- * fun main(args: Array<String>) {
- *   //sampleStart
- *   val result =
- *      mapOf("first" to ("A" to 1), "second" to ("B" to 2)).unzip()
- *   //sampleEnd
- *   println(result)
+ * fun test() {
+ *   mapOf(
+ *     "first" to ("A" to 1),
+ *     "second" to ("B" to 2)
+ *   ).unzip() shouldBe Pair(mapOf("first" to "A", "second" to "B"), mapOf("first" to 1, "second" to 2))
  * }
  * ```
  * <!--- KNIT example-map-07.kt -->
+ * <!--- lines.isEmpty() -->
  */
 public fun <K, A, B> Map<K, Pair<A, B>>.unzip(): Pair<Map<K, A>, Map<K, B>> =
-  entries.fold(emptyMap<K, A>() to emptyMap()) { (ls, rs), (k, v) ->
-    ls.plus(k to v.first) to rs.plus(k to v.second)
-  }
+  unzip { (_, pair) -> pair }
 
 /**
  * After applying the given function unzip the resulting structure into its elements.
  *
  * ```kotlin
  * import arrow.core.*
+ * import io.kotest.matchers.shouldBe
  *
- * fun main(args: Array<String>) {
- *   //sampleStart
- *   val result =
- *    mapOf("first" to "A:1", "second" to "B:2", "third" to "C:3").unzip { (_, e) ->
- *      e.split(":").let {
- *        it.first() to it.last()
- *      }
- *    }
- *   //sampleEnd
- *   println(result)
+ * fun test() {
+ *   mapOf("first" to "A:1", "second" to "B:2", "third" to "C:3").unzip { (_, e) ->
+ *     e.split(":").let {
+ *       it.first() to it.last()
+ *     }
+ *   } shouldBe Pair(
+ *     mapOf("first" to "A", "second" to "B", "third" to "C"),
+ *     mapOf("first" to "1", "second" to "2", "third" to "3")
+ *   )
  * }
  * ```
  * <!--- KNIT example-map-08.kt -->
+ * <!--- lines.isEmpty() -->
  */
-public fun <K, A, B, C> Map<K, C>.unzip(fc: (Map.Entry<K, C>) -> Pair<A, B>): Pair<Map<K, A>, Map<K, B>> =
-  mapValues(fc).unzip()
+public inline fun <K, A, B, C> Map<K, C>.unzip(fc: (Map.Entry<K, C>) -> Pair<A, B>): Pair<Map<K, A>, Map<K, B>> {
+  val lefts = mutableMapOf<K, A>()
+  val rights = mutableMapOf<K, B>()
+  forEach { entry ->
+    val (a, b) = fc(entry)
+    lefts[entry.key] = a
+    rights[entry.key] = b
+  }
+  return lefts to rights
+}
 
-public fun <K, V> Map<K, V>.getOrNone(key: K): Option<V> = this[key].toOption()
+@Suppress("UNCHECKED_CAST")
+public fun <K, V> Map<K, V>.getOrNone(key: K): Option<V> =
+  if (containsKey(key)) Some(get(key) as V) else None
```

Comment:
I'm wondering why you prefer `get(key) as V` as opposed to `get(key)!!`.

In fact, if we talk about style, I actually prefer something on the line of:
```suggestion
if (key in this) Some(this[key]) else None
```

Reply:
`this[key]` returns _nullable types_ and using `!!` results in `KotlinNullPointerException` with _nested null problem_ 😅 The original implementation contained a bug 😱